File: trees_and_graphs/count_good_nodes_in_binary_tree.js
/**
 * @param {TreeNode} root
 * @returns {number}
 */
var goodNodesRecursive = function(root)
{
    let dfs = (node, maxSoFar) => {
        if (!node)
        {
            return 0;
        }

        const left = dfs(node.left, Math.max(maxSoFar, node.val));
        const right = dfs(node.right, Math.max(maxSoFar, node.val));
        let ans = left + right;
        if (node.val >= maxSoFar)
        {
            ans++;
        }

        return ans;
    }

    return dfs(root, -Infinity);
};


var goodNodesIterative = function(root)
{
    if (!root)
    {
        return 0;
    }

    let stack = [[root, -Infinity]];
    let ans = 0;

    while (stack.length)
    {
        const [node, maxSoFar] = stack.pop();
        if (node.val >= maxSoFar)
        {
            ans++;
        }

        if (node.left)
        {
            stack.push([node.left, Math.max(maxSoFar, node.val)]);
        }

        if (node.right)
        {
            stack.push([node.right, Math.max(maxSoFar, node.val)]);
        }
    }

    return ans;
};

File: trees_and_graphs/count_good_nodes_in_binary_tree.py
import treenode

def goodNodesRecursive(root: treenode.TreeNode) -> int:
    def dfs(node, max_so_far):
        if not node:
            return 0
        
        left = dfs(node.left, max(max_so_far, node.val))
        right = dfs(node.right, max(max_so_far, node.val))
        ans = left + right
        if node.val >= max_so_far:
            ans += 1
        
        return ans
    
    return dfs(root, float("-inf"))

def goodNodesIterative(root: treenode.TreeNode) -> int:
    if not root:
        return 0
    
    stack = [(root, float("-inf"))]
    ans = 0

    while stack:
        node, max_so_far = stack.pop()
        if node.val >= max_so_far:
            ans += 1

        if node.left:
            stack.append((node.left, max(max_so_far, node.val)))
        
        if node.right:
            stack.append((node.right, max(max_so_far, node.val)))
    
    return ans